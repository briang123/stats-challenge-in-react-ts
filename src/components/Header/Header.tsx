import React from 'react';
import PropTypes from 'prop-types';
import logo from './../../logo.svg';
import styled from 'styled-components';
import { motion } from 'framer-motion';

const variants = {
  load: {
    scale: 2,
    transition: {
      type: 'spring',
    },
  },
  unload: {
    scale: 0.75,
  },
};

interface IProps {
  heading: string;
  description: string;
}

export const Header = ({ heading, description }: IProps) => {
  return (
    <StyledHeader data-testid="header">
      <Logo
        src={logo}
        alt="logo"
        variants={variants}
        initial="unload"
        animate="load"
        exit="unload"
      />
      <Heading data-testid="heading">{heading}</Heading>
      {description && <Description data-testid="description">{description}</Description>}
    </StyledHeader>
  );
};

Header.propTypes = {
  heading: PropTypes.string.isRequired,
  description: PropTypes.string,
};

Header.defaultProps = {
  heading: null,
  description: null,
};

export default Header;

const StyledHeader = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 20vh;
  font-size: calc(10px + 4vmin);
  background-color: var(--lightcolor);
  color: var(--darktext);
  margin-top: 25px;
  @media (min-width: 600px) {
    font-weight: bold;
    font-size: calc(10px + 2vmin);
  }
`;

const Logo = styled(motion.img)`
  height: 15vmin;
  pointer-events: none;
  @media (min-width: 600px) {
    height: 6vmin;
  }
`;

const Heading = styled.p`
  padding: 0;
  margin-bottom: 10px;
`;

const Description = styled.div`
  font-size: 0.5em;
  width: 90vw;
  color: var(--lighttext);
  margin-bottom: 10px;
  @media (min-width: 600px) {
    margin-bottom: 0px;
    width: 60vw;
  }
`;
