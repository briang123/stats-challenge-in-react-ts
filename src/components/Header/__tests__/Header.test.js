import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { create } from 'react-test-renderer';
import Header from '../Header';

afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();

describe('<Header />', () => {
  it('Tile component should fail without parameters', () => {
    render(<Header />);
    expect(console.error).toHaveBeenCalled();
  });

  it('Should match snapshot without heading text', () => {
    const tree = create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render logo and heading text', () => {
    const { getByTestId } = render(<Header heading="foo" />);

    const container = getByTestId('header');
    expect(container.tagName).toBe('DIV');
    expect(container.childElementCount).toBe(2);

    const logo = container.firstChild;
    expect(logo.tagName).toBe('IMG');

    const header = container.lastChild;
    expect(header.tagName).toBe('P');
    expect(header.textContent).toBe('foo');
  });

  it('Should match snapshot with heading text', () => {
    const tree = create(<Header heading="foo" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render logo, heading text, and description', () => {
    const { getByTestId, getByAltText } = render(
      <Header heading="foo" description="bar" />
    );

    const container = getByTestId('header');
    expect(container.tagName).toBe('DIV');
    expect(container.childElementCount).toBe(3);

    const logo = getByAltText('logo');
    expect(logo.tagName).toBe('IMG');

    const header = getByTestId('heading');
    expect(header.tagName).toBe('P');
    expect(header.textContent).toBe('foo');

    const description = getByTestId('description');
    expect(description.tagName).toBe('DIV');
    expect(description.textContent).toBe('bar');
  });

  it('Should match snapshot with heading and description text', () => {
    const tree = create(<Header heading="foo" descriptions="bar" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
