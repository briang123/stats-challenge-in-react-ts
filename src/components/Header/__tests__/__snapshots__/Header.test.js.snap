// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Should match snapshot with heading and description text 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-height: 20vh;
  font-size: calc(10px + 4vmin);
  background-color: var(--lightcolor);
  color: var(--darktext);
  margin-top: 25px;
}

.c1 {
  height: 15vmin;
  pointer-events: none;
}

.c2 {
  padding: 0;
  margin-bottom: 10px;
}

@media (min-width:600px) {
  .c0 {
    font-weight: bold;
    font-size: calc(10px + 2vmin);
  }
}

@media (min-width:600px) {
  .c1 {
    height: 6vmin;
  }
}

<div
  className="c0"
  data-testid="header"
>
  <img
    alt="logo"
    className="c1"
    src="logo.svg"
    style={
      Object {
        "transform": "scale(0.75) translateZ(0)",
      }
    }
  />
  <p
    className="c2"
    data-testid="heading"
  >
    foo
  </p>
</div>
`;

exports[`<Header /> Should match snapshot with heading text 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-height: 20vh;
  font-size: calc(10px + 4vmin);
  background-color: var(--lightcolor);
  color: var(--darktext);
  margin-top: 25px;
}

.c1 {
  height: 15vmin;
  pointer-events: none;
}

.c2 {
  padding: 0;
  margin-bottom: 10px;
}

@media (min-width:600px) {
  .c0 {
    font-weight: bold;
    font-size: calc(10px + 2vmin);
  }
}

@media (min-width:600px) {
  .c1 {
    height: 6vmin;
  }
}

<div
  className="c0"
  data-testid="header"
>
  <img
    alt="logo"
    className="c1"
    src="logo.svg"
    style={
      Object {
        "transform": "scale(0.75) translateZ(0)",
      }
    }
  />
  <p
    className="c2"
    data-testid="heading"
  >
    foo
  </p>
</div>
`;

exports[`<Header /> Should match snapshot without heading text 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-height: 20vh;
  font-size: calc(10px + 4vmin);
  background-color: var(--lightcolor);
  color: var(--darktext);
  margin-top: 25px;
}

.c1 {
  height: 15vmin;
  pointer-events: none;
}

.c2 {
  padding: 0;
  margin-bottom: 10px;
}

@media (min-width:600px) {
  .c0 {
    font-weight: bold;
    font-size: calc(10px + 2vmin);
  }
}

@media (min-width:600px) {
  .c1 {
    height: 6vmin;
  }
}

<div
  className="c0"
  data-testid="header"
>
  <img
    alt="logo"
    className="c1"
    src="logo.svg"
    style={
      Object {
        "transform": "scale(0.75) translateZ(0)",
      }
    }
  />
  <p
    className="c2"
    data-testid="heading"
  />
</div>
`;
